File: C++/two-furthest-houses-with-different-colors.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    int maxDistance(vector<int>& colors) {
        int result = 0;
        for (int i = 0; i < size(colors); ++i) {
            if (colors[i] != colors[0]) {
                result = max(result, i - 0);
            }
            if (colors[i] != colors.back()) {
                result = max(result, static_cast<int>(size(colors)) - 1 - i);
            }
        }
        return result;
    }
};

File: Python/two-furthest-houses-with-different-colors.py
# Time:  O(n)
# Space: O(1)

class Solution(object):
    def maxDistance(self, colors):
        """
        :type colors: List[int]
        :rtype: int
        """
        result = 0
        for i, x in enumerate(colors):
            if x != colors[0]:
                result = max(result, i)
            if x != colors[-1]:
                result = max(result, len(colors)-1-i)
        return result

